# Práctica 2: Bash y NGS

## Dulce María Montero Rasgado


## Parte I. Manejo de directorios

* **Creación de directorios**: Inicia la terminal y colócate en el directorio de tus documentos. Crea un directorio llamado: `p2_nombre_apellido`. Ingresa a la carpeta `p2_nombre_apellido` e imprime la ruta absoluta de ésta. 

Nota: No olivedes que debes ser reportados todos los **outputs**.

In [2]:
# Comando para crear directorio
mkdir

mkdir: falta un operando
Pruebe 'mkdir --help' para más información.


: 1

In [8]:
# Comando para ingresar a directorio
cd /home/dulma/Descargas/p2_Dulce_Montero

bash: cd: Descargas: No existe el fichero o el directorio


: 1

In [9]:
# Comando para imprimir ruta
pwd

/home/dulma/Descargas/p2_Dulce_Montero


: 1

* Indica que tipo de shell tiene tu computadora. Visita [el siguiente sitio](https://askubuntu.com/questions/590899/how-do-i-check-which-shell-i-am-using) para conocer el comando. 


In [1]:
# Comando para conocer shell
echo $0

/usr/bin/bash04l


: 1

* Un proyecto bioinformático que se quiere publicar, debe estar bien documentado y organizado. Dentro de tu carpeta de trabajo crea en una sola línea las siguientes directorios `data`, `raw_data`, `filtered`, `meta`, `scripts`, `figures`, `archive`. Después mueve `raw_data` y `filtered` dentro de `data`. 

In [10]:
# Comando para crear directorios
mkdir data raw_data filtered meta scripts figures archive

: 1

In [11]:
# Comando para mover raw_data y filtered
mv -t data raw_data filtered

: 1

* Visita [el siguiente repositorio](https://github.com/u-genoma/BioinfinvRepro/blob/master/Unidad2/Unidad2_Organizacion_proyecto_bioinf.md) de `GitHub` y contesta ¿A qué se debe el nombre y la organización de los directorios que acabamos de crear? 

## Parte II.  Manejo de archivos

En ésta parte trabajarás con secuencias genómicas de un organismo de tu interés (te sugiero que sea una bacteria, arquea o virus (excepto SARS-COV2) por el tamaño de los archivos). De ese organismo deberás conseguir: 

1. El archivo crudo (`fastq`). Para conseguir este archivo puedes consultar el base de datos de [SRA](https://www.ncbi.nlm.nih.gov/sra) o el [ENA](https://www.ebi.ac.uk/ena/browser/home). 
2. La secuencia de DNA (`fasta`).
3. El archivo con genes anotados (`gff`).
4. Un archivo con la descripción del proyecto, puede ser un Genbank o similar. 

* Guarda cada archivo con la siguiente estructura: `nombre_bicho.formato` (en `nombre_bicho` tienes que poner el nombre científico del organismo que elegiste).
* Descarga éstos cuatro archivos en tu carpeta `p2_nombre_apellido`.



* Cuéntame sobre el organismo que elegiste, elabora un resumen de 120 palabras de ese bicho.

_Bacillus cereus_ es una bacteria principalmente conocida por causar envenenamiento por consumo, es decir, intoxicaciones alimentarias que se manifiestan con síndrome emético y diarrea. 
El genoma completo de _B. cereus_ NC7401 comprende un único cromosoma circular de 5.221.581 pb, con un contenido de G+C del 35,6 %, es un genoma petite.
Además de su capacidad para provocar intoxicaciones alimentarias, B. cereus también está involucrada en infecciones locales y sistémicas, especialmente asociadas con pacientes inmunocomprometidos, neonatos, drogodependientes y pacientes con heridas quirúrgicas o traumáticas o catéteres. Las cepas aisladas de este tipo de infecciones han mostrado su habilidad para sintetizar exotoxinas necrotizantes semejantes a hemolisinas y fosfolipasas.
Se considera que su nicho fundamental es el suelo donde su gran diversidad metabólica le da ventaja sobre otros habitantes microbianos para obtener nutrientes.

* ¿Existe una publicación sobre la secuenciación y ensamblado de su genoma? Colóca la fuente en formato APA. 

Takeno, A., Okamoto, A., Tori, K., Oshima, K., Hirakawa, H., Toh, H., ... & Ohta, M. (2012). _Complete genome sequence of Bacillus cereus NC7401, which produces high levels of the emetic toxin cereulide._

* ¿Qué plataforma de secuenciación se utilizó? Explica brevemente las características de esa plataforma de secuenciación. 

Se utilizó Illumina NovaSeq; PacBio Sequel.
El sistema NovaSeq 6000 genera lecturas desde 80 Gb y 800 M de lecturas hasta 3 Tb y 10 B de lecturas en modo de celda de flujo único. En el modo de celda de doble flujo, el output de información puede ser hasta 6 Tb y 20 B de lecturas.

* Mueve tus cuatro archivos a la carpeta de `raw_data` y has un `ls` con formato largo y legible para humanos. Después copia el archivo `gff` a tu carpeta `filtered.`

In [14]:
# Comando para mover archivos
mv -t /home/dulma/Descargas/p2_Dulce_Montero/data/raw_data Bacillus_cereus.fasta Bacillus_cereus.gb Bacillus_cereus.gff3 Bacillus_cereus_fastq.fasta

: 1

In [16]:
# Accediendo a la carpeta raw_data
cd data/raw_data

: 1

In [26]:
# Comando para listado largo y legible para humanos
ls -lh

total 15M[?2004l
-rw-r--r-- 1 dulma dulma 5.2M may 13 18:58 Bacillus_cereus.fasta
-rw-r--r-- 1 dulma dulma 1.7M may 13 20:18 Bacillus_cereus.fastq
-rw-r--r-- 1 dulma dulma 5.5M may 13 19:00 Bacillus_cereus.gb
-rw-r--r-- 1 dulma dulma 2.5M may 13 19:02 Bacillus_cereus.gff3


: 1

In [19]:
# Comando para copiar archivo gff
cp Bacillus_cereus.gff3 ../filtered

: 1

* Explora tus tres archivos (`fastq`,`fasta`, y `gff`) , utiliza un comando para ver las primeras 12 líneas de cada archivo.

In [25]:
# Comando para ver fastq
head -n 12 Bacillus_cereus.fastq

@SRR10083550.1 0_0_1_12415_232/1
ACCAGAAATCCTATATACTAAATGGTTAATTGTAACTGAACCCGCAGGTCACATGTGTTAGGTTTCAC
+
GGGGGD=FGGADGGGGGGGGGFDDFGECCFGGGFGFEFFGDEFGGFGEDGGBEEGGDDFBGGGGGGFE
@SRR10083550.2 1_1_1_7227_342/1
TAGATATGTTTCGGGAGTTTACAGTTGCGTAGTTGCGGATTAAGGGAGTAGGATGAATATAACCCGCA
+
GGFDEFGGGGGGGDD@?AFGGAC?)BCBCCB@?EDBFF:,??B?=EDCCDGGGDD?DGFFGGGGFFGG
@SRR10083550.3 2_1_1_5884_292/1
CGAATACTGCATCCCAGCATGTCCTTTCTACTGCGGTAGCTCTGTCTTTTATAAAAATACTTTAGTCA
+
GFD>BB=?45@:CBEEEDDDFFFGFEGGEGG?GGGGGGFGGGGGGGGGEEFGBA=?EAEFD?77BABE


: 1

In [22]:
# Comando para ver fasta
head -n 12 Bacillus_cereus.fasta

>CP072774.1 Bacillus cereus strain BC33 chromosome, complete genome
TTGGAAAATATCTCTGATTTATGGAATAGTGCCTTAAAAGAATTAGAAAAAAAGGTAAGCAAGCCTAGTT
ATGAGACATGGTTAAAATCCACAACGGCTCATAACTTGAAGAAAGACGTATTAACGATTACAGCTCCGAA
TGAATTTGCTCGTGACTGGCTAGAATCTCATTACTCCGAACTAATTTCAGAAACACTATACGATTTAACA
GGGGCAAAATTAGCAATTCGCTTTATTATTCCCCAAAGTCAAGCTGAAGAGGACATTGATCTGCCTCCAG
TTAAGCCGAATCCAGCACAAGATGATTCAGCTCATTTACCACAGAGCATGTTAAATCCAAAATATACATT
CGATACATTTGTTATTGGCTCTGGTAACCGTTTTGCCCATGCAGCTTCATTAGCTGTAGCTGAGGCGCCA
GCTAAAGCGTATAATCCACTCTTTATTTACGGAGGAGTTGGACTTGGAAAGACACATTTAATGCATGCAA
TTGGTCATTATGTAATTGAACATAATCCAAATGCAAAAGTTGTATATTTATCATCAGAAAAATTTACAAA
TGAATTTATTAACTCTATTCGTGATAATAAAGCTGTTGATTTTCGTAATAAATATCGCAACGTAGATGTT
TTATTGATAGATGATATTCAATTTCTTGCTGGAAAAGAACAGACTCAAGAAGAGTTTTTCCATACATTTA
ATGCATTACACGAAGAAAGTAAACAAATTGTAATTTCTAGTGACCGACCACCAAAAGAAATTCCAACTTT


: 1

In [24]:
# Comando para ver gff
head -n 12 Bacillus_cereus.gff3

##sequence-region CP072774.1 1 5325341
##species https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=1396
CP072774.1	Genbank	region	1	5325341	.	+	.	ID=CP072774.1:1..5325341;Dbxref=taxon:1396;Is_circular=true;Name=ANONYMOUS;collection-date=2020-07-01;country=China: Shanghai;gbkey=Src;genome=chromosome;isolation-source=Bee Honey;mol_type=genomic DNA;strain=BC33
CP072774.1	Genbank	gene	1	1341	.	+	.	ID=gene-J8Y18_00005;Name=dnaA;gbkey=Gene;gene=dnaA;gene_biotype=protein_coding;locus_tag=J8Y18_00005
CP072774.1	Protein Homology	CDS	1	1341	.	+	0	ID=cds-QUW37220.1;Parent=gene-J8Y18_00005;Dbxref=NCBI_GP:QUW37220.1;Name=QUW37220.1;gbkey=CDS;gene=dnaA;inference=COORDINATES: similar to AA sequence:RefSeq:NP_387882.1;locus_tag=J8Y18_00005;product=chromosomal replication initiator protein DnaA;protein_id=QUW37220.1;transl_table=11
CP072774.1	Genbank	gene	1520	2665	.	+	.	ID=gene-J8Y18_00010;Name=dnaN;gbkey=Gene;gene=dnaN;gene_biotype=protein_coding;locus_tag=J8Y18_00010
CP072774.1	Protein Hom

: 1

* Ya que los exploraste, identifica un caracter en los archivos `fastq` y `fasta` que te ayuden a contar el número de secuencias que contiene cada uno. Ahora, utilliza un comando para conocer el número de secuencias en cada archivo. 

In [27]:
# Comando para contar secuencias en fastq
grep '@' -c Bacillus_cereus.fastq

1576204h


: 1

In [28]:
# Comando para contar secuencias en fasta
grep '>' -c Bacillus_cereus.fasta

# explota pero es una forma de contar el numero de secuencias xdxd

1[?2004h


: 1

* Explica la diferencia entre los formatos `faa`, `fna`, `fastq`, `fasta` y `gff`

- El formato .faa es utilizada por el NCBI para "FASTA Amino Acids", por lo tanto, es igualmente un archivo FASTA pero destinado a secuencias de aminoacidos.
- .fna es un formato FASTA estandarizado para datos de DNA And Protein Sequence Alignment.FNA.
- .fastq es un formato de archivo que integra lecturas provenientes de secuenciación, y la puntuación de la calidad, asociada a dichas lecturas, base por base. 
- .gff en inglés gene-finding format o generic feature format, es un archivo con extensión “.gff” creado con el objetivo de identificar y describir genes, sus características y secuencias proteicas. 

* Abre el archivo `gff` de las siguientes formas `less archivo.gff3` y `less -S archivo.gff3`. Observa las diferencias y explicalas. 

In [ ]:
# less archivo.gff3
less Bacillus_cereus.gff3

##sequence-region CP072774.1 1 5325341
##species https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=1396
CP072774.1      Genbank region  1       5325341 .       +       .       ID=CP072774.1:1..5325341;Dbxref=taxon:1396;Is_circular=true;Name=ANONYMOUS;collection-date=2020-07-01;country=China: Shanghai;gbkey=Src;genome=chromosome;isolation-source=Bee Honey;mol_type=genomic DNA;strain=BC33
CP072774.1      Genbank gene    1       1341    .       +       .       ID=gene-J8Y18_00005;Name=dnaA;gbkey=Gene;gene=dnaA;gene_biotype=protein_coding;locus_tag=J8Y18_00005
CP072774.1      Protein Homology        CDS     1       1341    .       +       0       ID=cds-QUW37220.1;Parent=gene-J8Y18_00005;Dbxref=NCBI_GP:QUW37220.1;Name=QUW37220.1;gbkey=CDS;gene=dnaA;inference=COORDINATES: similar to AA sequence:RefSeq:NP_387882.1;locus_tag=J8Y18_00005;product=chromosomal replication initiator protein DnaA;protein_id=QUW37220.1;transl_table=11
CP072774.1      Genbank gene    1520    2665    .      

In [ ]:
# less -S archivo.gff3
less -S Bacillus_cereus.gff3

#### Así se ve, no sé por qué no imprimió la vista del comando, pero en mi consola sale esto

##sequence-region CP072774.1 1 5325341
##species https://www.ncbi.nlm.nih.gov/Taxonomy/Browser/wwwtax.cgi?id=1396
CP072774.1      Genbank region  1       5325341 .       +       .       ID=CP072774.1:1..5325341;Dbx>
CP072774.1      Genbank gene    1       1341    .       +       .       ID=gene-J8Y18_00005;Name=dna>
CP072774.1      Protein Homology        CDS     1       1341    .       +       0       ID=cds-QUW37>
CP072774.1      Genbank gene    1520    2665    .       +       .       ID=gene-J8Y18_00010;Name=dna>
CP072774.1      Protein Homology        CDS     1520    2665    .       +       0       ID=cds-QUW37>
CP072774.1      Genbank gene    2793    3005    .       +       .       ID=gene-J8Y18_00015;Name=yaa>
CP072774.1      Protein Homology        CDS     2793    3005    .       +       0       ID=cds-QUW37>
CP072774.1      Genbank gene    3018    4145    .       +       .       ID=gene-J8Y18_00020;Name=rec>
CP072774.1      Protein Homology        CDS     3018    4145    .       +       0       ID=cds-QUW37>
CP072774.1      Genbank gene    4184    6106    .       +       .       ID=gene-J8Y18_00025;Name=gyr>
CP072774.1      Protein Homology        CDS     4184    6106    .       +       0       ID=cds-QUW37>
CP072774.1      Genbank gene    6195    8666    .       +       .       ID=gene-J8Y18_00030;Name=gyr>
CP072774.1      Protein Homology        CDS     6195    8666    .       +       0       ID=cds-QUW37>
CP072774.1      Genbank gene    8908    10459   .       +       .       ID=gene-J8Y18_00035;Name=J8Y>
CP072774.1      cmsearch        rRNA    8908    10459   .       +       .       ID=rna-J8Y18_00035;P>
CP072774.1      cmsearch        exon    8908    10459   .       +       .       ID=exon-J8Y18_00035->
CP072774.1      Genbank gene    10590   10666   .       +       .       ID=gene-J8Y18_00040;Name=J8Y>
CP072774.1      tRNAscan-SE     tRNA    10590   10666   .       +       .       ID=rna-J8Y18_00040;P>
CP072774.1      tRNAscan-SE     exon    10590   10666   .       +       .       ID=exon-J8Y18_00040->

* Entra a tu carpeta de `filtered` utilizando una ruta relativa. 

In [ ]:
# Comando
cd ../filtered

* Cambia el nombre del archivo que tienes aquí por `cp_nombre_bicho.gff` (en `nombre_bicho` tienes que poner el nombre científico del organismo que elegiste).

In [ ]:
# Comando para cambiar nombre de un archivo
mv Bacillus_cereus.gff3 cp_Bacillus_cereus.gff3

* Filtra el tercer campo de tu archivo `gff` por la categoría `gene` y reporta la cantidad de genes que tiene el archivo. Contesta: 

1. ¿A qué corresponde el campo tres? 
2. ¿Cuál es la diferencia entre gene y CDS?

In [ ]:
# Comando para concer cuántos genes hay
grep 'gene' -c  cp_Bacillus_cereus.gff3
# 10822

In [ ]:
cut -f 3 cp_Bacillus_cereus.gff3 

El campo tres revela el tipo de transcrito que hay en los 10882 genes. Puede ser un gen, tRNA, rRNA, tmRNA...
CDS es la region de codificacion de un gen, es la parte del gen que codifica un transcrito, en cambio gen tiene a toda la secuencia completa incluyendo intrones.

## Parte III. Visualización con Python

* En tu archivo `gff`, identifica cuántas categorías distintas existen en el campo 3 y cuántas veces aparecen. Redirecciona la salida a un archivo que se llame `barplot_data.txt`

In [ ]:
# Comando
awk 'BEGIN{print "Transcrito:"}; {tipo[$3] +=1}; END {for (i in tipo) {if (i != "") {print i "\t" tipo[i]}}}' data/filtered/cp_Bacillus_cereus.gff3 > barplot_data.txt

Transcrito:
ncRNA	2
SRP_RNA	1
sequence_feature	3
tmRNA	1
exon	153
rRNA	42
region	1
CDS	21
riboswitch	37
Homology	5258
RNase_P_RNA	1
binding_site	43
gene	5346
pseudogene	85
tRNA	106
1	1

* Crea una figura a partir de `barplot_data.txt`, para esto puedes utiliza el siguiete script de Python para generar una figura. 

```python

import numpy as np
import matplotlib.pyplot as plt
from textwrap import wrap

# Crea el dataset con tus datos obtenidos en barplot_data.txt
frecuencias = [3, 12, 5, 18, 45]
# Cambia el nombre de las categorías por las que obtuviste no lo dejes como "Categoría A", etc.
categorias = ['Categoría A', 'Categoría B', 'Categoría C', 'Categoría D', 'Categoría E']
categorias = [ ''.join(wrap(l, 11)) for l in categorias]

y_pos = np.arange(len(categorias))

# Gráfico de barras
plt.bar(y_pos, frecuencias)

# Nombres en el eje-x
plt.xticks(y_pos, categorias)

# Mostrar la gráfica
plt.show()

```

También puedes generar tu figura con R o Julia, **prohibido** usar Excel.

* Guarda la figura en tu carpeta `figures` con el nombre `barplot.png`.
* Guardar el código que utilizaste en la carpeta `scripts`, lo puedes guardar en cualquier formato. 

* Guarda tu libreta de `Jupyter` o tu `md` con tu nombre, las preguntas, los comandos y los outputs en la carpeta `scripts`. 

* Comprime tu carpeta `p2_nombre_apellido` y súbela en la sección de Classroom correspondiente.

In [1]:
ls mv -t figures 

 archive       data      meta                    'Parte III.ipynb'
 barplot.png   figures   p2_Montero_Dulce.ipynb   scripts


: 1